File: prolog/src/graphics.pl
%% ============================================================================
%% RADARPAS Prolog Translation - Graphics Module
%% ============================================================================
%% Translation of EGA graphics routines from RADAR.PAS
%% Models the graphics operations as declarative transformations on a
%% framebuffer state, since Prolog cannot do direct hardware I/O.
%%
%% The original code directly manipulates EGA video memory at segment A000h
%% using plane selection, bit masking, and function selection registers.
%%
%% Original: D. G. Lane, January 14, 1988
%% ============================================================================

:- module(graphics, [
    % Graphics state
    make_graphics_state/1,
    graphics_plane/2,
    graphics_func/2,
    graphics_cursor/3,
    graphics_char_size/2,
    graphics_window/5,

    % Plane operations
    select_plane/3,
    show_plane/3,

    % Function selection
    select_func/3,

    % Mask operations
    set_mask/3,

    % Cursor operations
    goto_xy/4,

    % Plotting primitives
    gr_plot/4,
    gr_line/6,

    % Text rendering
    gr_write_str/5,

    % Screen operations
    clear_screen/2,
    draw_scale/2,
    init_ega/2,

    % Window management
    open_window/6,
    un_window/2,

    % Message display
    gr_message/4,

    % Display line (RLE decompression)
    disp_line/4,

    % Toggle graphics
    toggle_graphics/2,

    % Graphics function types
    valid_func/1
]).

:- use_module(types).

%% ============================================================================
%% Graphics Function Types
%% ============================================================================
%% FuncType = (Rot1..Rot7, _Clr, _And, _Or, _Xor)
%% EGA ALU function selection

valid_func(rot1).
valid_func(rot2).
valid_func(rot3).
valid_func(rot4).
valid_func(rot5).
valid_func(rot6).
valid_func(rot7).
valid_func(clr).
valid_func('and').
valid_func('or').
valid_func('xor').

%% ============================================================================
%% Graphics State Structure
%% ============================================================================
%% gfx_state(Plane, Func, Mask, CursorX, CursorY,
%%           XPos, YPos, XMax, YMax, CharSize,
%%           GraphicsOn, Framebuffer)
%%
%% Plane:      Current selected EGA planes (list of 0..3)
%% Func:       Current ALU function (clr/and/or/xor/rotN)
%% Mask:       Current bit mask (0x00..0xFF)
%% CursorX/Y:  Text cursor position
%% XPos/YPos:  Window origin offset
%% XMax/YMax:  Window extent
%% CharSize:   Character height (8 or 14)
%% GraphicsOn: Whether graphics planes are visible
%% Framebuffer: Abstract representation of video memory

make_graphics_state(
    gfx_state([2], clr, 0xFF, 0, 0,
              0, 0, 79, 24, 14,
              true, [])
).

graphics_plane(gfx_state(P, _, _, _, _, _, _, _, _, _, _, _), P).
graphics_func(gfx_state(_, F, _, _, _, _, _, _, _, _, _, _), F).
graphics_cursor(gfx_state(_, _, _, X, Y, _, _, _, _, _, _, _), X, Y).
graphics_char_size(gfx_state(_, _, _, _, _, _, _, _, _, CS, _, _), CS).
graphics_window(gfx_state(_, _, _, _, _, XP, YP, XM, YM, _, _, _), XP, YP, XM, YM).

%% ============================================================================
%% Plane Operations
%% ============================================================================

%% select_plane(+Planes, +StateIn, -StateOut)
%% Select EGA planes for write operations
%% Original: port[$3C4]:=$02; port[$3C5]:=Data
select_plane(Planes,
    gfx_state(_, Func, Mask, CX, CY, XP, YP, XM, YM, CS, GO, FB),
    gfx_state(Planes, Func, Mask, CX, CY, XP, YP, XM, YM, CS, GO, FB)
).

%% show_plane(+Planes, +StateIn, -StateOut)
%% Select EGA planes for display (attribute controller)
%% Original: port[$3C0]:=$12; port[$3C0]:=Data; port[$3C0]:=$20
show_plane(Planes,
    gfx_state(P, Func, Mask, CX, CY, XP, YP, XM, YM, CS, _, FB),
    gfx_state(P, Func, Mask, CX, CY, XP, YP, XM, YM, CS, GO, FB)
) :-
    (   Planes = [0, 1, 2, 3] -> GO = true
    ;   Planes = [0, 1]       -> GO = false
    ;   GO = true  % default: show all
    ).

%% ============================================================================
%% Function Selection
%% ============================================================================

%% select_func(+Func, +StateIn, -StateOut)
%% Select EGA ALU function for bitwise operations
%% Original: port[$3CE]:=$03; port[$3CF]:=encoded_func
select_func(Func,
    gfx_state(P, _, Mask, CX, CY, XP, YP, XM, YM, CS, GO, FB),
    gfx_state(P, Func, Mask, CX, CY, XP, YP, XM, YM, CS, GO, FB)
) :-
    valid_func(Func).

%% ============================================================================
%% Mask Operations
%% ============================================================================

%% set_mask(+MaskValue, +StateIn, -StateOut)
%% Set bit mask register for pixel operations
%% Original: port[$3CE]:=$08; port[$3CF]:=ToMask
set_mask(MaskVal,
    gfx_state(P, Func, _, CX, CY, XP, YP, XM, YM, CS, GO, FB),
    gfx_state(P, Func, MaskVal, CX, CY, XP, YP, XM, YM, CS, GO, FB)
).

%% ============================================================================
%% Cursor Operations
%% ============================================================================

%% goto_xy(+X, +Y, +StateIn, -StateOut)
%% Position text cursor
goto_xy(X, Y,
    gfx_state(P, Func, Mask, _, _, XP, YP, XM, YM, CS, GO, FB),
    gfx_state(P, Func, Mask, X, Y, XP, YP, XM, YM, CS, GO, FB)
).

%% ============================================================================
%% Plotting Primitives
%% ============================================================================

%% gr_plot(+X, +Y, +StateIn, -StateOut)
%% Plot single pixel at (X, Y)
%% Original: AtByte:=Y*80+(X shr 3); AtBit:=$80 shr (X and $07)
gr_plot(X, Y,
    gfx_state(P, Func, _, CX, CY, XP, YP, XM, YM, CS, GO, FB),
    gfx_state(P, Func, 0xFF, CX, CY, XP, YP, XM, YM, CS, GO, NewFB)
) :-
    % Calculate byte offset and bit position
    AtByte is Y * 80 + (X >> 3),
    AtBit is 0x80 >> (X /\ 0x07),
    % Add pixel to framebuffer as operation record
    append(FB, [plot(AtByte, AtBit, P, Func)], NewFB).

%% gr_line(+X1, +Y1, +X2, +Y2, +StateIn, -StateOut)
%% Draw line from (X1,Y1) to (X2,Y2) using Bresenham-style algorithm
%% Original: incremental DDA line drawing
gr_line(X1, Y1, X2, Y2, StateIn, StateOut) :-
    % Record line drawing operation
    StateIn = gfx_state(P, Func, Mask, CX, CY, XP, YP, XM, YM, CS, GO, FB),
    append(FB, [line(X1, Y1, X2, Y2, P, Func)], NewFB),
    StateOut = gfx_state(P, Func, Mask, CX, CY, XP, YP, XM, YM, CS, GO, NewFB).

%% ============================================================================
%% Text Rendering
%% ============================================================================

%% gr_write_str(+String, +X, +Y, +StateIn, -StateOut)
%% Write string at pixel coordinates using 8-pixel-wide character table
%% Original: procedure GRWrite with bit-shift blitting
gr_write_str(String, X, Y, StateIn, StateOut) :-
    StateIn = gfx_state(P, Func, Mask, CX, CY, XP, YP, XM, YM, CS, GO, FB),
    append(FB, [text(String, X, Y, P, Func)], NewFB),
    StateOut = gfx_state(P, Func, Mask, CX, CY, XP, YP, XM, YM, CS, GO, NewFB).

%% ============================================================================
%% Screen Operations
%% ============================================================================

%% clear_screen(+StateIn, -StateOut)
%% Clear radar display area (circular region)
%% Original: uses Circle1 table to determine extent at each scanline
clear_screen(
    gfx_state(P, Func, Mask, CX, CY, XP, YP, XM, YM, CS, GO, _),
    gfx_state(P, Func, Mask, CX, CY, XP, YP, XM, YM, CS, GO, [clear])
).

%% draw_scale(+StateIn, -StateOut)
%% Draw range scale along left edge
%% Original: vertical lines with tick marks at 10, 50, 100 pixel intervals
draw_scale(StateIn, StateOut) :-
    StateIn = gfx_state(P, _, Mask, CX, CY, XP, YP, XM, YM, CS, GO, FB),
    append(FB, [scale], NewFB),
    StateOut = gfx_state(P, clr, Mask, CX, CY, XP, YP, XM, YM, CS, GO, NewFB).

%% init_ega(+StateIn, -StateOut)
%% Initialize EGA 640x350 16-color mode
%% Original: BIOS INT 10h, AH=00, AL=10h
init_ega(_, State) :-
    make_graphics_state(State).

%% ============================================================================
%% Window Management
%% ============================================================================

%% open_window(+X, +Y, +XSize, +YSize, +StateIn, -StateOut)
%% Open a text window at character position (X,Y) with given size
%% Original: procedure Window - fills area with blue background
open_window(X, Y, XSize, YSize,
    gfx_state(P, Func, Mask, _, _, _, _, _, _, CS, GO, FB),
    gfx_state(P, Func, Mask, 0, 0, X, Y, XSize, YSize, CS, GO, NewFB)
) :-
    append(FB, [window(X, Y, XSize, YSize)], NewFB).

%% un_window(+StateIn, -StateOut)
%% Close window and restore full-screen display
%% Original: procedure UnWindow - clears and redraws overlays
un_window(
    gfx_state(P, Func, Mask, _, _, _, _, _, _, CS, GO, FB),
    gfx_state(P, Func, Mask, 0, 0, 0, 0, 79, 24, CS, GO, NewFB)
) :-
    append(FB, [un_window], NewFB).

%% ============================================================================
%% Message Display
%% ============================================================================

%% gr_message(+Text, +WaitForKey, +StateIn, -StateOut)
%% Display centered message at bottom of screen
%% Original: procedure GRMessage - opens temporary window at row 24
gr_message(Text, WaitForKey, StateIn, StateOut) :-
    StateIn = gfx_state(P, Func, Mask, CX, CY, XP, YP, XM, YM, CS, GO, FB),
    append(FB, [message(Text, WaitForKey)], NewFB),
    StateOut = gfx_state(P, Func, Mask, CX, CY, XP, YP, XM, YM, CS, GO, NewFB).

%% ============================================================================
%% Run-Length Decompression (DispLine)
%% ============================================================================

%% disp_line(+CompressedData, +StateIn, -StateOut, -LineNum)
%% Decompress and display one radar scan line
%% Original: procedure DispLine
%%
%% Compressed format: pairs of (control, size) bytes
%%   Control byte bits 5-6: color (0=none, 1=red, 2=green, 3=both)
%%   Control byte bits 0-2 + size byte: run length (11-bit)
%%   Terminator: 0x18 (Cancel character)
%%
%% In Prolog, we decode the RLE data into a list of run segments

disp_line(CompressedData, StateIn, StateOut, LineNum) :-
    % First two bytes are the line number (div 54)
    CompressedData = [LoByte, HiByte | RunData],
    RawLine is LoByte + (HiByte << 8),
    LineNum is RawLine // 54,

    % Decode runs
    decode_runs(RunData, Runs),

    % Record display operation
    StateIn = gfx_state(P, _, Mask, CX, CY, XP, YP, XM, YM, CS, GO, FB),
    append(FB, [radar_line(LineNum, Runs)], NewFB),
    StateOut = gfx_state(P, clr, Mask, CX, CY, XP, YP, XM, YM, CS, GO, NewFB).

%% decode_runs(+Bytes, -Runs)
%% Decode run-length encoded byte pairs into run segments
%% Each pair: [ControlByte, SizeByte]
%% Run = run(Color, Length)
decode_runs([], []).
decode_runs([0x18 | _], []).     % Terminator
decode_runs([Control, Size | Rest], [run(Color, Length) | MoreRuns]) :-
    Control \= 0x18,
    % Decode color from bits 5-6
    ColorBits is (Control >> 5) /\ 3,
    decode_color(ColorBits, Color),
    % Decode length from bits 0-2 of control + full size byte
    Length is ((Control /\ 0x07) << 8) \/ Size,
    decode_runs(Rest, MoreRuns).

decode_color(0, none).
decode_color(1, red).
decode_color(2, green).
decode_color(3, both).

%% ============================================================================
%% Toggle Graphics
%% ============================================================================

%% toggle_graphics(+StateIn, -StateOut)
%% Toggle between radar image only and full overlay display
%% Original: if GraphicsOn then ShowPlane([0,1]) else ShowPlane([0..3])
toggle_graphics(
    gfx_state(P, Func, Mask, CX, CY, XP, YP, XM, YM, CS, true, FB),
    gfx_state(P, Func, Mask, CX, CY, XP, YP, XM, YM, CS, false, FB)
).
toggle_graphics(
    gfx_state(P, Func, Mask, CX, CY, XP, YP, XM, YM, CS, false, FB),
    gfx_state(P, Func, Mask, CX, CY, XP, YP, XM, YM, CS, true, FB)
).

%% ============================================================================
%% END OF MODULE
%% ============================================================================

File: elixir/lib/radarpas/graphics.ex
defmodule Radarpas.Graphics do
  @moduledoc """
  Graphics module - Drawing primitives and screen management.
  Translated from the Graphics Routines section of RADAR.PAS (lines 181-508).

  The original code directly manipulated EGA video memory at segment $A000
  using plane selection via ports $3C4/$3C5 and function selection via $3CE/$3CF.
  This translation abstracts the rendering into drawing commands that can be
  executed by any renderer (terminal, GUI, etc.).

  Original procedures translated:
    SelectPlane, ShowPlane, SelectFunc, SetMask, GotoXY, ConOut, ReadStr,
    GRWrite, GRPlot, GRLine, Window, GRMessage, Ask, ToggleGraphics,
    DispLine, DrawScale, InitEGA
  """

  alias Radarpas.CoreTypes.TimeRec

  # ============================================================================
  # Graphics State
  # Original global variables: CurrPlane, CurrFunc, CurrMask, CursX, CursY,
  #   XPos, YPos, XMax, YMax, CharSize, GraphicsOn, LastMess
  # ============================================================================

  defstruct curr_plane: MapSet.new([0, 1, 2, 3]),
            curr_func: :clear,
            curr_mask: 0xFF,
            curs_x: 0,
            curs_y: 0,
            x_pos: 0,
            y_pos: 0,
            x_max: 79,
            y_max: 24,
            char_size: 14,
            graphics_on: true,
            last_mess: ""

  @type plane_set :: MapSet.t()

  # Original: FuncType = (Rot1,Rot2,Rot3,Rot4,Rot5,Rot6,Rot7,_Clr,_And,_Or,_Xor)
  @type func_type :: :rot1 | :rot2 | :rot3 | :rot4 | :rot5 | :rot6 | :rot7 |
                     :clear | :band | :bor | :bxor

  @type draw_command ::
          {:set_pixel, x :: integer(), y :: integer()}
          | {:draw_line, x1 :: integer(), y1 :: integer(), x2 :: integer(), y2 :: integer()}
          | {:fill_rect, x :: integer(), y :: integer(), width :: integer(), height :: integer(),
             value :: byte()}
          | {:draw_text, x :: integer(), y :: integer(), text :: String.t()}
          | {:clear_area, x :: integer(), y :: integer(), width :: integer(), height :: integer()}
          | {:set_planes, plane_set()}
          | {:set_func, func_type()}

  # ============================================================================
  # State Transformations
  # ============================================================================

  @doc """
  Select active write plane(s).
  Original: procedure SelectPlane(ForPlanes: PlaneSet) - lines 199-204
  In the original, this wrote directly to EGA sequencer port $3C4/$3C5.
  """
  def select_plane(%__MODULE__{} = state, planes) do
    %{state | curr_plane: MapSet.new(List.wrap(planes))}
  end

  @doc """
  Select graphics function (write mode).
  Original: procedure SelectFunc(ForFunc: FuncType) - lines 219-225
  Wrote to EGA graphics controller port $3CE/$3CF.
  """
  def select_func(%__MODULE__{} = state, func) do
    %{state | curr_func: func}
  end

  @doc """
  Set bit mask for pixel operations.
  Original: procedure SetMask(ToMask: byte) - lines 228-232
  """
  def set_mask(%__MODULE__{} = state, mask) do
    %{state | curr_mask: mask}
  end

  @doc """
  Move cursor to position.
  Original: procedure GotoXY(X,Y: integer) - lines 241-244
  """
  def goto_xy(%__MODULE__{} = state, x, y) do
    %{state | curs_x: x, curs_y: y}
  end

  @doc """
  Set up a text window region.
  Original: procedure Window(X,Y,XSize,YSize: byte) - lines 350-363
  Cleared the window area in EGA memory and set position/size limits.
  """
  def window(%__MODULE__{} = state, x, y, x_size, y_size) do
    %{state | x_pos: x, y_pos: y, x_max: x_size, y_max: y_size, curs_x: 0, curs_y: 0}
  end

  @doc """
  Toggle between text and graphics display mode.
  Original: procedure ToggleGraphics - lines 401-405
  Toggled EGA planes [0,1] vs [0..3] via ShowPlane.
  """
  def toggle_graphics(%__MODULE__{} = state) do
    %{state | graphics_on: not state.graphics_on}
  end

  # ============================================================================
  # Drawing Commands (Pure - generate command lists)
  # ============================================================================

  @doc """
  Plot a single pixel.
  Original: procedure GRPlot(X,Y: integer) - lines 311-322
  Calculated byte position as Y*80+(X shr 3), bit as $80 shr (X and $07),
  then read-modify-wrote to EGA memory at $A000.
  """
  def plot(x, y) when is_integer(x) and is_integer(y) do
    {:set_pixel, x, y}
  end

  @doc """
  Draw a line between two points.
  Original: procedure GRLine(X1,Y1,X2,Y2: integer) - lines 324-348
  Used a simple DDA (Digital Differential Analyzer) algorithm:
  - If |dy| > |dx|: step along Y, interpolate X as real
  - Otherwise: step along X, interpolate Y as real
  """
  def line(x1, y1, x2, y2) do
    {:draw_line, x1, y1, x2, y2}
  end

  @doc """
  Generate the individual pixel commands for a line (DDA algorithm).
  This is a faithful translation of the original GRLine procedure.
  """
  def line_points(x1, y1, x2, y2) do
    dx = x2 - x1
    dy = y2 - y1

    if abs(dy) > abs(dx) do
      # Step along Y axis
      add = dx / abs(dy)
      dir = if y2 > y1, do: 1, else: -1
      do_line_y(x1 + 0.0, y1, y2, add, dir, [])
    else
      # Step along X axis
      add = if x2 != x1, do: dy / abs(dx), else: 0.0
      dir = if x2 > x1, do: 1, else: -1
      do_line_x(x1, y1 + 0.0, x2, add, dir, [])
    end
  end

  defp do_line_y(x_real, y1, y2, _add, _dir, acc) when y1 == y2, do: Enum.reverse(acc)

  defp do_line_y(x_real, y1, y2, add, dir, acc) do
    point = {:set_pixel, trunc(x_real), y1}
    do_line_y(x_real + add, y1 + dir, y2, add, dir, [point | acc])
  end

  defp do_line_x(x1, y_real, x2, _add, _dir, acc) when x1 == x2, do: Enum.reverse(acc)

  defp do_line_x(x1, y_real, x2, add, dir, acc) do
    point = {:set_pixel, x1, trunc(y_real)}
    do_line_x(x1 + dir, y_real + add, x2, add, dir, [point | acc])
  end

  @doc """
  Write text at graphics position.
  Original: procedure GRWrite(ForStr: linetype; X,Y: integer) - lines 296-309
  Rendered characters from the ROM character table (CharTab8) into EGA memory
  with bit-shift rotation for sub-byte X alignment.
  """
  def gr_write(text, x, y) do
    {:draw_text, x, y, text}
  end

  @doc """
  Display a message at the bottom of screen (line 24).
  Original: procedure GRMessage(ForStr: linetype; WaitKey: boolean) - lines 366-393
  Centered the message, clearing previous message first.
  """
  def gr_message(text) do
    x = max(0, 40 - div(byte_size(text), 2))
    {:draw_text, x, 24, text}
  end

  @doc """
  Draw the vertical scale bar on the left edge.
  Original: procedure DrawScale - lines 465-478
  Drew a scale ruler from row 0 to 349, with markers at 10/50/100-pixel intervals.
  """
  def draw_scale do
    for i <- 0..349 do
      width =
        cond do
          rem(i, 100) == 0 -> 8
          rem(i, 50) == 0 -> 6
          rem(i, 10) == 0 -> 4
          true -> 2
        end

      {:fill_rect, 0, i, width, 1, 0xFF}
    end
  end

  @doc """
  Clear the circular radar display area.
  Original: procedure ClearScreen - lines 783-792
  Cleared within the circle boundary defined by Circle1 lookup table.
  """
  def clear_screen(circle1) do
    for i <- 0..175 do
      half_width = 1 + Bitwise.bsr(Enum.at(circle1, i), 3)
      x = 40 - half_width

      [
        {:clear_area, x, i, half_width * 2, 1},
        {:clear_area, x, 350 - i, half_width * 2, 1}
      ]
    end
    |> List.flatten()
  end

  @doc """
  Decompress and display a radar picture line.
  Original: procedure DispLine(var LinePtr) - lines 410-463
  The picture data was run-length encoded with 2-byte entries:
    bits 15-13: color (0=off, 1=red, 2=green, 3=yellow)
    bits 10-0: pixel count
  Terminated by $18 byte. Line number stored in first 2 bytes.
  """
  def disp_line(line_data) when is_binary(line_data) do
    <<line_num::little-16, rest::binary>> = line_data
    line_num = div(line_num, 54)
    {commands, _} = decode_line_segments(rest, line_num * 80, 0, [])
    {line_num, commands}
  end

  defp decode_line_segments(<<0x18, _rest::binary>>, _at_byte, _at_bit, acc) do
    {Enum.reverse(acc), :done}
  end

  defp decode_line_segments(<<hi, lo, rest::binary>>, at_byte, at_bit, acc) do
    size = Bitwise.bsl(Bitwise.band(hi, 0x07), 8) ||| lo

    color =
      case Bitwise.band(Bitwise.bsr(hi, 5), 0x03) do
        0 -> :off
        1 -> :red
        2 -> :green
        3 -> :yellow
      end

    # Generate pixel commands for this segment
    new_cmds =
      if color != :off do
        for pixel <- 0..(size - 1) do
          bit_pos = at_bit + pixel
          x = (at_byte + div(bit_pos, 8)) * 8 + rem(bit_pos, 8)
          y_offset = div(at_byte, 80)
          {:set_pixel, rem(x, 640), y_offset}
        end
      else
        []
      end

    new_at_bit = at_bit + size
    new_at_byte = at_byte + div(new_at_bit, 8)
    new_at_bit = rem(new_at_bit, 8)

    decode_line_segments(rest, new_at_byte, new_at_bit, new_cmds ++ acc)
  end

  defp decode_line_segments(<<>>, _at_byte, _at_bit, acc) do
    {Enum.reverse(acc), :truncated}
  end

  @doc """
  Initialize EGA mode (640x350 16-color).
  Original: procedure InitEGA - lines 480-508
  Set video mode $10 via BIOS INT 10h, loaded character tables,
  initialized palette registers, and drew initial parameter labels.
  Returns initial draw commands.
  """
  def init_ega(%TimeRec{} = time) do
    [
      {:clear_area, 0, 0, 640, 350},
      {:set_planes, MapSet.new([2])},
      {:draw_text, 0, 0, "TILT  : XXX"},
      {:draw_text, 65, 0, "TIME : " <> format_time(time)},
      {:draw_text, 0, 24, "RANGE : XXX"},
      {:draw_text, 65, 24, "GAIN  : XXX"},
      {:draw_text, 0, 22, "H = HELP"}
    ]
  end

  defp format_time(%TimeRec{hour: h, minute: m}) do
    h_str = if h < 13, do: "#{h}", else: "#{h - 12}"
    m_str = String.pad_leading("#{m}", 2, "0")
    suffix = if h > 12, do: "pm", else: "am"
    "#{String.pad_leading(h_str, 2)}:#{m_str}#{suffix}"
  end

  # ============================================================================
  # Renderer Behaviour
  # Original rendering went directly to EGA memory ($A000 segment).
  # This behaviour allows pluggable renderers.
  # ============================================================================

  @callback execute(draw_command()) :: :ok
  @callback execute_all([draw_command()]) :: :ok
  @callback present() :: :ok
end
